enhancements/gateway/20220130-distinguishing-central-agent.md

```diff
@@ -0,0 +1,88 @@
+
+Instead of relying on a persistent label to identify a cluster, it might be better to use runtime-only information that is not stored persistently, but instead is an attribute of an agent's current session. This has some advantages:
+- Because distinguishing an agent is primarily intended to be used for *authorization* purposes, if the agent isn't connected to the gateway, it wouldn't need to be authorized to do anything anyway. The usefulness of knowing which agent *was* the central agent seems limited.
+- This eliminates the possibility that configuration issues could cause the central agent to be misidentified, or that the central agent could be impersonated by another agent.
+- This removes the need to keep track of the label and handle cases where it is removed or changed
+- This removes the need to expand on either the bootstrap or token management APIs.
+
+Therefore, the best course of action might be to have the distinguishing label be a property of an agent's status. The current status message looks like this:
+
+```protobuf
+message Status {
+  google.protobuf.Timestamp timestamp = 1;
+  bool connected = 2;
+}
+```
+
+We could add a new field containing session attributes:
+
+```protobuf
+message Status {
+  google.protobuf.Timestamp timestamp = 1;
+  bool connected = 2;
+  map<string, string> sessionAttributes = 3;
+}
+```
+
+This field could contain implementation-specific and/or well-known labels that can be used for a variety of purposes, including distinguishing the central agent.
+
+Implementing session attributes is far simpler by contrast:
+1. The agent adds an additional header to its connection request, containing a list of session attributes it wants to assert.
+2. For each attribute, the gateway sends an additional challenge to the agent during the authentication handshake. The agent must then compute a MAC for the challenge using an implementation-specific secret, and send the result back to the gateway.
+3. The gateway verifies the MACs and adds the attributes to the agent's status if they are valid. If any of the attributes are invalid, the agent is disconnected.
+
+An example implementation of the implementation-specific secret for the central agent could be:
+1. An octet key pair is generated by cert-manager and mounted into both the agent and gateway pods
+2. The location of this keypair on disk is configured in both the agent and gateway config file (or checked for in a well-known location)
+3. The presence of the keypair triggers the agent to assert the relevant attribute.
+
+Configuration could be either done via config files or environment variables, or both.
+
+## Acceptance criteria:
+- [ ] The central agent can be distinguished from other agents using a well-known label in its session attributes list, obtained from the management API.
+- [ ] An agent with an ID that is not equal to the gateway's ID can successfully be set as the central agent.
+- [ ] Two agents can simultaneously be labeled as central agents.
```

Comment:
What is the use case for having two agents as central agents?  Wouldn't we expect there to be one and only one?

Reply:
In this context "two agents" just means two agents with the same connection attribute. Could happen if we need to scale the agent or something similar.